lib/Tools/LinearLayout.cpp

```diff
@@ -655,19 +656,99 @@ LinearLayout::divideRight(const LinearLayout &divisor) {
                          divisor.getInDimSizeLog2(inDim));
   }
 
+  // Record size 1 out-dims caused by the division.
   llvm::MapVector<StringAttr, int32_t> newOutDims = outDims;
-  for (const auto [outDim, outDimSize] : divisor.outDims) {
-    if (newOutDims[outDim] < outDimSize) {
+  llvm::DenseSet<size_t> sizeOneOutDimIndices;
+  for (auto [outDimName, outDimSize] : divisor.outDims) {
+    if (newOutDims[outDimName] < outDimSize) {
       return std::nullopt;
     }
-    newOutDims[outDim] /= outDimSize;
+    auto newOutDimSize = newOutDims[outDimName] / outDimSize;
+    if (newOutDimSize == 1) {
+      sizeOneOutDimIndices.insert(getOutDimIndex(outDimName));
+    }
+    newOutDims[outDimName] = newOutDimSize;
   }
 
-  eraseEmptyInOutDims(newBases, newOutDims);
+  // Only erase the trailing empty out-dims.
+  // For example, consider l = o / r, where:
+  //
+  //   out-dims(o) = ["out0", "out1", "out2", "out3"]
+  //   out-dims(r) = ["out1", "out3"]
+  //
+  // If we remove "out1" from o, we get:
+  //
+  //   out-dims(l) = ["out0", "out2", "out3"]
+  //
+  // Then, out-dims(l * r) = ["out0", "out2", "out3"] * ["out1", "out3"] =
+  // ["out0", "out2", "out3", "out1"] which does not match out-dims(o).
+  //
+  // However, if we only remove "out3", we get:
+  //
+  //   out-dims(l) = ["out0", "out1", "out2"]
+  //
+  // Then, out-dims(l * r) = ["out0", "out1", "out2"] * ["out1", "out3"] =
+  // ["out0", "out1", "out2", "out3"] which matches out-dims(o).
+  llvm::SmallVector<size_t> emptyOutDimIndices;
+  for (int outDimIdx = outDims.size() - 1; outDimIdx >= 0; outDimIdx--) {
+    if (sizeOneOutDimIndices.contains(outDimIdx)) {
+      emptyOutDimIndices.push_back(outDimIdx);
+      newOutDims.erase(newOutDims.begin() + outDimIdx);
+    } else {
+      break;
+    }
+  }
+
+  // Erase the basis elements corresponding to the empty out-dims.
+  for (auto &[inDim, inDimBases] : newBases) {
+    for (auto &basis : inDimBases) {
+      for (int i : emptyOutDimIndices) {
+        basis.erase(basis.begin() + i);
+      }
+    }
+  }
+
+  // In some cases, we may have input dimensions that map to all zeros for
+  // certain output dimensions.  For example, suppose the current layout is:
+  //
+  //              out-dim0  out-dim1
+  //   in-dim0 |      0        0
+  //   in-dim1 |      0        1
+  //   in-dim2 |      0        0
+  //
+  // And the layout of the divisor is:
+  //
+  //              out-dim1
+  //   in-dim1 |      1
+  //
+  // The quotient should be:
+  //
+  //              out-dim0
+  //   in-dim0 |   size 1
+  //   in-dim1 |   size 1
+  //   in-dim2 |   size 1
```

Comment:
I don't really know what you mean with these diagrams.  Perhaps we could use something that matches the toString() output of LinearLayout, since at least that has a well-defined meaning?

Reply:
Updated